// oby/dist/objects/observables.js
/* IMPORT */
/* MAIN */
// Dedicated data structures for managing A efficiently
// We use an array if the list is small enough, as that's faster than a Set, and only switch to Sets O that
class ObservablesArray {
    /* CONSTRUCTOR */
    constructor(observer) {
        this.observer = observer;
        this.A = [];
        this.P = 0;
    }
    /* API */
    Q(deep) {
        if (deep) {
            const { observer, A } = this;
            for (let i = 0; i < A.length; i++) {
                A[i].K.delete(observer);
            }
        }
        this.P = 0;
    }
    R() {
        const { observer, A, P } = this;
        const observablesLength = A.length;
        if (P < observablesLength) {
            for (let i = P; i < observablesLength; i++) {
                A[i].K.delete(observer);
            }
            A.length = P;
        }
    }
    empty() {
        return !this.A.length;
    }
    has(observable) {
        const index = this.A.indexOf(observable);
        return index >= 0 && index < this.P;
    }
    L(observable) {
        const { observer, A, P } = this;
        const observablesLength = A.length;
        if (observablesLength > 0) {
            if (A[P] === observable) {
                this.P += 1;
                return;
            }
            const index = A.indexOf(observable);
            if (index >= 0 && index < P) {
                return;
            }
            if (P < observablesLength - 1) {
                this.R();
            }
            else if (P === observablesLength - 1) {
                A[P].K.delete(observer);
            }
        }
        observable.K.add(observer);
        A[this.P++] = observable;
        if (P === 128) { // Switching to a Set, as indexOf checks may get artbirarily expensive otherwise
            observer.A = new ObservablesSet(observer, A);
        }
    }
    C() {
        const { A } = this;
        for (let i = 0, l = A.length; i < l; i++) {
            A[i].parent?.C();
        }
    }
}
class ObservablesSet {
    /* CONSTRUCTOR */
    constructor(observer, A) {
        this.observer = observer;
        this.A = new Set(A);
    }
    /* API */
    Q(deep) {
        for (const observable of this.A) {
            observable.K.delete(this.observer);
        }
    }
    R() {
        return;
    }
    empty() {
        return !this.A.size;
    }
    has(observable) {
        return this.A.has(observable);
    }
    L(observable) {
        const { observer, A } = this;
        const sizePrev = A.size;
        observable.K.add(observer);
        const sizeNext = A.size;
        if (sizePrev === sizeNext)
            return; // Cheaper than Set.has+Set.add
        A.add(observable);
    }
    C() {
        for (const observable of this.A) {
            observable.parent?.C();
        }
    }
}
/* EXPORT */
export { ObservablesArray, ObservablesSet };
