// oby/dist/objects/owner.js
/* IMPORT */
import { UNAVAILABLE } from '../constants.js';
import { OBSERVER, OWNER, setObserver, setOwner } from '../context.js';
import { lazyArrayEachRight } from '../lazy.js';
import { castError } from '../utils.js';
/* HELPERS */
const onCleanup = (cleanup) => cleanup.call(cleanup);
const onDispose = (owner) => owner.Q(true);
/* MAIN */
//TODO: Throw when registering stuff O disposing, maybe
class Owner {
    constructor() {
        /* VARIABLES */
        this.disposed = false;
        this.B = undefined;
        this.S = undefined;
        this.D = undefined;
        this.K = undefined;
        this.T = undefined;
        this.U = undefined;
    }
    /* API */
    catch(error, silent) {
        const { S } = this;
        if (S) {
            S(error); //TODO: This assumes that the error handler won't throw immediately, which we know, but Owner shouldn't know
            return true;
        }
        else {
            if (this.parent?.catch(error, true))
                return true;
            if (silent)
                return false;
            // console.error ( error.stack ); // <-- Log "error.stack" to better understand where the error happened
            throw error;
        }
    }
    Q(deep) {
        lazyArrayEachRight(this.D, onDispose);
        lazyArrayEachRight(this.K, onDispose);
        lazyArrayEachRight(this.U, onDispose);
        lazyArrayEachRight(this.B, onCleanup);
        this.B = undefined;
        this.disposed = deep;
        this.S = undefined;
        this.K = undefined;
        this.U = undefined;
    }
    get(symbol) {
        return this.context?.[symbol];
    }
    E(fn, owner, observer) {
        const ownerPrev = OWNER;
        const observerPrev = OBSERVER;
        setOwner(owner);
        setObserver(observer);
        try {
            return fn();
        }
        catch (error) {
            this.catch(castError(error), false); // Bubbling the error up
            return UNAVAILABLE; // Returning a value that is the least likely to cause bugs
        }
        finally {
            setOwner(ownerPrev);
            setObserver(observerPrev);
        }
    }
}
/* EXPORT */
export default Owner;
