// oby/dist/objects/suspense.js
/* IMPORT */
import { DIRTY_MAYBE_YES, DIRTY_YES } from '../constants.js';
import { OWNER } from '../context.js';
import { lazyArrayEach, lazyArrayPush, lazySetEach } from '../lazy.js';
import Effect from './effect.js';
import Owner from './owner.js';
import { SYMBOL_SUSPENSE } from '../symbols.js';
import { isFunction } from '../utils.js';
/* MAIN */
class Suspense extends Owner {
    /* CONSTRUCTOR */
    constructor() {
        super();
        /* VARIABLES */
        this.parent = OWNER;
        this.context = { ...OWNER.context, [SYMBOL_SUSPENSE]: this };
        lazyArrayPush(this.parent, 'U', this);
        this.suspended = (OWNER.get(SYMBOL_SUSPENSE)?.suspended || 0);
    }
    /* API */
    A4(force) {
        if (!this.suspended && !force)
            return; // Already suspended, this can happen at instantion time
        const suspendedPrev = this.suspended;
        const suspendedNext = suspendedPrev + (force ? 1 : -1);
        this.suspended = suspendedNext;
        if (!!suspendedPrev === !!suspendedNext)
            return; // Same state, nothing to pause or resume
        /* NOTIFYING OBSERVERS, ROOTS AND SUSPENSES */
        this.observable?.set(!!suspendedNext);
        const notifyOwner = (owner) => {
            lazyArrayEach(owner.D, notifyOwner);
            lazyArrayEach(owner.K, notifyObserver);
            lazyArrayEach(owner.U, notifySuspense);
            lazySetEach(owner.T, notifyRoot);
        };
        const notifyObserver = (observer) => {
            if (observer instanceof Effect) {
                if (observer.J === DIRTY_MAYBE_YES || observer.J === DIRTY_YES) {
                    if (observer.init) {
                        observer.C();
                    }
                    else {
                        observer.F();
                    }
                }
            }
            notifyOwner(observer);
        };
        const notifyRoot = (root) => {
            if (isFunction(root)) {
                root().forEach(notifyOwner);
            }
            else {
                notifyOwner(root);
            }
        };
        const notifySuspense = (suspense) => {
            suspense.A4(force);
        };
        notifyOwner(this);
    }
    E(fn) {
        return super.E(fn, this, undefined);
    }
}
/* EXPORT */
export default Suspense;
