// oby@0.54.0/dist/index.js
//TODO: Explore making setting A inside effects and memos more explicit/strict
//TODO: Explore using Solid's double-array optimization to delete the Set
//TODO: Support auto-disposing of effects and memos (free up some memory)
//TODO: Support frozen A more deeply again
//TODO: Support faster B based on root disposal
//TODO: Support pooling in keyed $.for
/* IMPORT */
import $ from './methods/$.js';
import batch from './methods/batch.js';
import boolean from './methods/boolean.js';
import cleanup from './methods/cleanup.js';
import context from './methods/context.js';
import disposed from './methods/disposed.js';
import effect from './methods/effect.js';
import _for from './methods/for.js';
import get from './methods/get.js';
import _if from './methods/if.js';
import isBatching from './methods/is_batching.js';
import isObservable from './methods/is_observable.js';
import isStore from './methods/is_store.js';
import memo from './methods/memo.js';
import observable from './methods/observable.js';
import owner from './methods/owner.js';
import readonly from './methods/readonly.js';
import resolve from './methods/resolve.js';
import root from './methods/root.js';
import selector from './methods/selector.js';
import store from './methods/store.js';
import suspended from './methods/suspended.js';
import suspense from './methods/suspense.js';
import _switch from './methods/switch.js';
import ternary from './methods/ternary.js';
import tick from './methods/tick.js';
import tryCatch from './methods/try_catch.js';
import untrack from './methods/untrack.js';
import untracked from './methods/untracked.js';
import _with from './methods/with.js';
import { SYMBOL_OBSERVABLE, SYMBOL_OBSERVABLE_FROZEN, SYMBOL_OBSERVABLE_READABLE, SYMBOL_OBSERVABLE_WRITABLE, SYMBOL_STORE, SYMBOL_STORE_KEYS, SYMBOL_STORE_OBSERVABLE, SYMBOL_STORE_TARGET, SYMBOL_STORE_VALUES, SYMBOL_UNCACHED, SYMBOL_UNTRACKED, SYMBOL_UNTRACKED_UNWRAPPED } from './symbols.js';
/* EXPORT */
export default $;
export { batch, boolean, cleanup, context, disposed, effect, _for as for, get, _if as if, isBatching, isObservable, isStore, memo, observable, owner, readonly, resolve, root, selector, store, suspended, suspense, _switch as switch, ternary, tick, tryCatch, untrack, untracked, _with as with };
export { SYMBOL_OBSERVABLE, SYMBOL_OBSERVABLE_FROZEN, SYMBOL_OBSERVABLE_READABLE, SYMBOL_OBSERVABLE_WRITABLE, SYMBOL_STORE, SYMBOL_STORE_KEYS, SYMBOL_STORE_OBSERVABLE, SYMBOL_STORE_TARGET, SYMBOL_STORE_VALUES, SYMBOL_UNCACHED, SYMBOL_UNTRACKED, SYMBOL_UNTRACKED_UNWRAPPED };
