// oby@14.3.5/dist/objects/scheduler.async.js
/* IMPORT */
import { BATCH } from '../context.js';
/* MAIN */
// Using 2 microtasks to give a chance to things using 1 microtask (like refs in Voby) to run first
class Scheduler {
    constructor() {
        /* VARIABLES */
        this.A1 = [];
        this.A2 = false;
        this.A3 = false;
        /* QUEUING API */
        this.N = () => {
            if (this.A2)
                return;
            if (!this.A1.length)
                return;
            try {
                this.A2 = true;
                while (true) {
                    const queue = this.A1;
                    if (!queue.length)
                        break;
                    this.A1 = [];
                    for (let i = 0, l = queue.length; i < l; i++) {
                        queue[i].C();
                    }
                }
            }
            finally {
                this.A2 = false;
            }
        };
        this.queue = () => {
            if (this.A3)
                return;
            this.A3 = true;
            this.resolve();
        };
        this.resolve = () => {
            queueMicrotask(() => {
                queueMicrotask(() => {
                    if (BATCH) {
                        BATCH.finally(this.resolve);
                    }
                    else {
                        this.A3 = false;
                        this.N();
                    }
                });
            });
        };
        /* SCHEDULING API */
        this.F = (effect) => {
            this.A1.push(effect);
            this.queue();
        };
    }
}
/* EXPORT */
export default new Scheduler();
