// flac-stream-tagger/dist/metadata-block/MetadataBlockHeader.js
import { BufferBase, allocBufferAndWrite } from "../lib/BufferBase.js";
export var MetadataBlockType;
(function (MetadataBlockType) {
    MetadataBlockType[MetadataBlockType["StreamInfo"] = 0] = "StreamInfo";
    MetadataBlockType[MetadataBlockType["Padding"] = 1] = "Padding";
    MetadataBlockType[MetadataBlockType["Application"] = 2] = "Application";
    MetadataBlockType[MetadataBlockType["SeekTable"] = 3] = "SeekTable";
    MetadataBlockType[MetadataBlockType["VorbisComment"] = 4] = "VorbisComment";
    MetadataBlockType[MetadataBlockType["CueSheet"] = 5] = "CueSheet";
    MetadataBlockType[MetadataBlockType["Picture"] = 6] = "Picture";
    MetadataBlockType[MetadataBlockType["Invalid"] = 127] = "Invalid";
})(MetadataBlockType || (MetadataBlockType = {}));
export const MetadataBlockHeaderLength = 4;
export class MetadataBlockHeader extends BufferBase {
    type;
    dataLength;
    isLast;
    constructor(type, dataLength = 0, isLast = false) {
        super();
        this.type = type;
        this.dataLength = dataLength;
        this.isLast = isLast;
        this.isLast = isLast;
        this.type = type;
        this.dataLength = dataLength;
    }
    static SIZE = 4;
    static fromBuffer(buffer) {
        const lastAndType = buffer.readUint8();
        return new MetadataBlockHeader((lastAndType & 0b01111111), buffer.readUintBE(1, 3), (lastAndType & 0b10000000) === 1);
    }
    toBuffer() {
        return allocBufferAndWrite(this.length, (buffer) => {
            buffer.writeUint8(this.type + (this.isLast ? 0b10000000 : 0));
            buffer.writeUintBE(this.dataLength, 1, 3);
        });
    }
    get length() {
        return MetadataBlockHeaderLength;
    }
}
