// flac-stream-tagger/dist/metadata-block/PictureBlock.js
import imageinfo from "imageinfo";
import { MetadataBlockHeader, MetadataBlockHeaderLength, MetadataBlockType } from "./MetadataBlockHeader.js";
import { MetadataBlock } from "./MetadataBlock.js";
import { allocBufferAndWrite } from "../lib/BufferBase.js";
export var PictureType;
(function (PictureType) {
    PictureType[PictureType["Other"] = 0] = "Other";
    PictureType[PictureType["FileIcon"] = 1] = "FileIcon";
    PictureType[PictureType["OtherFileIcon"] = 2] = "OtherFileIcon";
    PictureType[PictureType["FrontCover"] = 3] = "FrontCover";
    PictureType[PictureType["BackCover"] = 4] = "BackCover";
    PictureType[PictureType["LeafletPage"] = 5] = "LeafletPage";
    PictureType[PictureType["Media"] = 6] = "Media";
    PictureType[PictureType["LeadArtist"] = 7] = "LeadArtist";
    PictureType[PictureType["Artist"] = 8] = "Artist";
    PictureType[PictureType["Conductor"] = 9] = "Conductor";
    PictureType[PictureType["Band"] = 10] = "Band";
    PictureType[PictureType["Composer"] = 11] = "Composer";
    PictureType[PictureType["Lyricist"] = 12] = "Lyricist";
    PictureType[PictureType["RecordingLocation"] = 13] = "RecordingLocation";
    PictureType[PictureType["DuringRecording"] = 14] = "DuringRecording";
    PictureType[PictureType["DuringPerformance"] = 15] = "DuringPerformance";
    PictureType[PictureType["MovieScreenCapture"] = 16] = "MovieScreenCapture";
    // eslint-disable-next-line @typescript-eslint/naming-convention
    PictureType[PictureType["ABrightColouredFish"] = 17] = "ABrightColouredFish";
    PictureType[PictureType["Illustration"] = 18] = "Illustration";
    PictureType[PictureType["BandLogotype"] = 19] = "BandLogotype";
    PictureType[PictureType["PublisherLogotype"] = 20] = "PublisherLogotype";
})(PictureType || (PictureType = {}));
export class PictureBlock extends MetadataBlock {
    header;
    pictureType;
    mime;
    description;
    width;
    height;
    colorDepth;
    colors;
    pictureBuffer;
    constructor(initialValues) {
        super();
        const { header = new MetadataBlockHeader(MetadataBlockType.Picture), pictureType = PictureType.FrontCover, mime, description = "", width, height, colorDepth = 24, colors = 0, buffer, } = initialValues;
        this.header = header;
        this.pictureType = pictureType;
        if (mime && width && height) {
            this.mime = mime;
            this.width = width;
            this.height = height;
        }
        else {
            const info = imageinfo(buffer);
            this.mime = mime ?? info.mimeType;
            this.width = width ?? info.width;
            this.height = height ?? info.height;
        }
        this.description = description;
        this.colorDepth = colorDepth;
        this.colors = colors;
        this.pictureBuffer = buffer;
    }
    static fromBuffer(buffer) {
        let bufferIndex = 0;
        const header = MetadataBlockHeader.fromBuffer(buffer);
        if (header.type !== MetadataBlockType.Picture)
            throw new Error(`Invalid picture block header type! Expected type ${MetadataBlockType.Picture} got ${header.type}`);
        bufferIndex += header.length;
        const pictureType = buffer.readUintBE(bufferIndex, 4);
        bufferIndex += 4;
        const mimeLength = buffer.readUintBE(bufferIndex, 4);
        bufferIndex += 4;
        const mime = buffer.subarray(bufferIndex, bufferIndex + mimeLength).toString();
        bufferIndex += mimeLength;
        const descriptionLength = buffer.readUintBE(bufferIndex, 4);
        bufferIndex += 4;
        const description = buffer.subarray(bufferIndex, bufferIndex + descriptionLength).toString();
        bufferIndex += descriptionLength;
        const width = buffer.readUintBE(bufferIndex, 4);
        bufferIndex += 4;
        const height = buffer.readUintBE(bufferIndex, 4);
        bufferIndex += 4;
        const colorDepth = buffer.readUintBE(bufferIndex, 4);
        bufferIndex += 4;
        const colors = buffer.readUintBE(bufferIndex, 4);
        bufferIndex += 4;
        const pictureDataLength = buffer.readUintBE(bufferIndex, 4);
        bufferIndex += 4;
        const pictureBuffer = buffer.subarray(bufferIndex, bufferIndex + pictureDataLength);
        return new PictureBlock({
            header,
            pictureType,
            mime,
            description,
            width,
            height,
            colorDepth,
            colors,
            buffer: pictureBuffer,
        });
    }
    toBuffer() {
        return Buffer.concat([
            this.header.toBuffer(),
            allocBufferAndWrite(4, (b) => b.writeUint32BE(this.pictureType)),
            allocBufferAndWrite(4, (b) => b.writeUint32BE(Buffer.byteLength(this.mime))),
            Buffer.from(this.mime),
            allocBufferAndWrite(4, (b) => b.writeUint32BE(Buffer.byteLength(this.description))),
            Buffer.from(this.description),
            allocBufferAndWrite(4, (b) => b.writeUint32BE(this.width)),
            allocBufferAndWrite(4, (b) => b.writeUint32BE(this.height)),
            allocBufferAndWrite(4, (b) => b.writeUint32BE(this.colorDepth)),
            allocBufferAndWrite(4, (b) => b.writeUint32BE(this.colors)),
            allocBufferAndWrite(4, (b) => b.writeUint32BE(this.pictureBuffer.length)),
            this.pictureBuffer,
        ]);
    }
    toPicture() {
        return {
            pictureType: this.pictureType,
            mime: this.mime,
            description: this.description,
            colorDepth: this.colorDepth,
            colors: this.colors,
            buffer: this.pictureBuffer,
        };
    }
    get length() {
        return (MetadataBlockHeaderLength +
            4 + // type length
            4 + // mime length
            Buffer.byteLength(this.mime) +
            4 + // description length
            Buffer.byteLength(this.description) +
            4 * 5 + // width, height, color depth, colors, picture data length
            this.pictureBuffer.length);
    }
}
