// oby@0.54.0/dist/methods/for.cache.keyed.js
/* IMPORT */
import { OWNER } from '../context.js';
import { lazySetAdd, lazySetDelete } from '../lazy.js';
import cleanup from './cleanup.js';
import resolve from './resolve.js';
import { frozen, readable } from '../objects/callable.js';
import Observable from '../objects/observable.js';
import Root from '../objects/root.js';
import { SYMBOL_CACHED, SYMBOL_SUSPENSE, SYMBOL_UNCACHED } from '../symbols.js';
/* HELPERS */
const DUMMY_INDEX = frozen(-1);
class MappedRoot extends Root {
}
/* MAIN */
class CacheKeyed {
    /* CONSTRUCTOR */
    constructor(fn) {
        /* VARIABLES */
        this.parent = OWNER;
        this.suspense = OWNER.get(SYMBOL_SUSPENSE);
        this.A5 = new Map();
        this.A6 = false; // The A6 is flipped with each iteration, the T that don't have the updated one are disposed, it's like a cheap M basically
        this.A7 = 0; // Number of previous items
        this.A8 = 0; // Number of previous items that got reused
        this.A9 = 0; // Number of next items
        /* API */
        this.cleanup = () => {
            if (!this.A7)
                return; // There was nothing AA, no need to cleanup
            if (this.A7 === this.A8)
                return; // All the previous items were reused, no need to cleanup
            const { A5, A6 } = this;
            if (!A5.size)
                return; // Nothing to Q of
            if (this.A9) { // Regular cleanup
                A5.forEach((mapped, value) => {
                    if (mapped.A6 === A6)
                        return;
                    mapped.Q(true);
                    A5.delete(value);
                });
            }
            else { // There is nothing O, disposing quickly
                this.A5.forEach(mapped => {
                    mapped.Q(true);
                });
                this.A5 = new Map();
            }
        };
        this.Q = () => {
            if (this.suspense) {
                lazySetDelete(this.parent, 'T', this.T);
            }
            this.A7 = this.A5.size;
            this.A8 = 0;
            this.A9 = 0;
            this.cleanup();
        };
        this.AA = () => {
            this.A6 = !this.A6;
            this.A8 = 0;
            this.A9 = 0;
        };
        this.O = (values) => {
            this.A9 = values.length;
            this.cleanup();
            this.A7 = this.A9;
            this.A8 = 0;
        };
        this.map = (values) => {
            this.AA();
            const { A5, A6, fn, AB } = this;
            const results = new Array(values.length);
            let resultsCached = true; // Whether all results are cached, if so this enables an optimization
            let resultsUncached = true; // Whether all results are anew, if so this enables an optimization in Voby
            let A8 = 0;
            for (let i = 0, l = values.length; i < l; i++) {
                const value = values[i];
                const cached = A5.get(value);
                if (cached && cached.A6 !== A6) {
                    resultsUncached = false;
                    A8 += 1;
                    cached.A6 = A6;
                    cached.index?.set(i);
                    results[i] = cached.G; //TSC
                }
                else {
                    resultsCached = false;
                    const mapped = new MappedRoot(false);
                    if (cached) {
                        cleanup(() => mapped.Q(true));
                    }
                    mapped.E(() => {
                        let index = DUMMY_INDEX;
                        if (AB) {
                            mapped.index = new Observable(i);
                            index = readable(mapped.index);
                        }
                        const G = results[i] = resolve(fn(value, index));
                        mapped.A6 = A6;
                        mapped.G = G;
                        if (!cached) {
                            A5.set(value, mapped);
                        }
                    });
                }
            }
            this.A8 = A8;
            this.O(values);
            if (resultsCached) {
                results[SYMBOL_CACHED] = true;
            }
            if (resultsUncached) {
                results[SYMBOL_UNCACHED] = true;
            }
            return results;
        };
        this.T = () => {
            return Array.from(this.A5.values());
        };
        this.fn = fn;
        this.AB = (fn.length > 1);
        if (this.suspense) {
            lazySetAdd(this.parent, 'T', this.T);
        }
    }
}
/* EXPORT */
export default CacheKeyed;
