// musicbrainz-api/lib/httpClient.js
import { CookieJar } from "tough-cookie";
export class HttpClient {
    constructor(options) {
        this.options = options;
        this.cookieJar = new CookieJar();
    }
    get(path, options) {
        return this._fetch('get', path, options);
    }
    post(path, options) {
        return this._fetch('post', path, options);
    }
    postForm(path, formData, options) {
        const encodedFormData = new URLSearchParams(formData).toString();
        return this._fetch('post', path, { ...options, body: encodedFormData, headers: { 'Content-Type': 'application/x-www-form-urlencoded' } });
    }
    // biome-ignore lint/complexity/noBannedTypes:
    postJson(path, json, options) {
        const encodedJson = JSON.stringify(json);
        return this._fetch('post', path, { ...options, body: encodedJson, headers: { 'Content-Type': 'application/json.' } });
    }
    async _fetch(method, path, options) {
        if (!options)
            options = {};
        let url = `${this.options.baseUrl}/${path}`;
        if (options.query) {
            url += `?${new URLSearchParams(options.query)}`;
        }
        const cookies = await this.getCookies();
        const headers = {
            ...options.headers,
            'User-Agent': this.options.userAgent,
            'Cookie': cookies
        };
        const response = await fetch(url, {
            method,
            ...options,
            headers,
            body: options.body,
            redirect: options.followRedirects === false ? 'manual' : 'follow'
        });
        await this.registerCookies(response);
        return response;
    }
    registerCookies(response) {
        const cookie = response.headers.get('set-cookie');
        if (cookie) {
            return this.cookieJar.setCookie(cookie, response.url);
        }
    }
    getCookies() {
        return this.cookieJar.getCookieString(this.options.baseUrl); // Get cookies for the request
    }
}
