// oby@0.54.0/dist/objects/effect.js
/* IMPORT */
import { lazyArrayPush } from '../lazy.js';
import Observer from './observer.js';
import Scheduler from './scheduler.async.js';
import { SYMBOL_SUSPENSE } from '../symbols.js';
import { isFunction } from '../utils.js';
/* MAIN */
class Effect extends Observer {
    /* CONSTRUCTOR */
    constructor(fn, options) {
        super();
        this.fn = fn;
        if (options?.suspense !== false) {
            const suspense = this.get(SYMBOL_SUSPENSE);
            if (suspense) {
                this.suspense = suspense;
            }
        }
        if (options?.sync === true) {
            this.sync = true;
        }
        if (options?.sync === 'init') {
            this.init = true;
            this.C();
        }
        else {
            this.F();
        }
    }
    /* API */
    run() {
        const G = super.H(this.fn);
        if (isFunction(G)) {
            lazyArrayPush(this, 'B', G);
        }
    }
    F() {
        if (this.suspense?.suspended)
            return;
        if (this.sync) {
            this.C();
        }
        else {
            Scheduler.F(this);
        }
    }
    I(J) {
        const statusPrev = this.J;
        if (statusPrev >= J)
            return;
        this.J = J;
        if (!this.sync || (statusPrev !== 2 && statusPrev !== 3)) { // It isn't currently executing, so let's F it
            this.F();
        }
    }
    C() {
        if (this.suspense?.suspended)
            return;
        super.C();
    }
}
/* EXPORT */
export default Effect;
