// oby@0.54.0/dist/objects/observer.js
/* IMPORT */
import { DIRTY_NO, DIRTY_MAYBE_NO, DIRTY_MAYBE_YES, DIRTY_YES } from '../constants.js';
import { OWNER, SUPER_OWNER } from '../context.js';
import { lazyArrayPush } from '../lazy.js';
import { ObservablesArray, ObservablesSet } from './observables.js';
import Owner from './owner.js';
/* MAIN */
class Observer extends Owner {
    /* CONSTRUCTOR */
    constructor() {
        super();
        /* VARIABLES */
        this.parent = OWNER;
        this.context = OWNER.context;
        this.J = DIRTY_YES;
        this.A = new ObservablesArray(this);
        if (OWNER !== SUPER_OWNER) {
            lazyArrayPush(this.parent, 'K', this);
        }
    }
    /* API */
    Q(deep) {
        this.A.Q(deep);
        super.Q(deep);
    }
    H(fn) {
        this.Q(false);
        this.J = DIRTY_MAYBE_NO; // Resetting the trip flag, we didn't re-execute just yet
        try {
            return this.E(fn, this, this);
        }
        finally {
            this.A.R();
        }
    }
    run() {
        throw new Error('Abstract method');
    }
    I(J) {
        throw new Error('Abstract method');
    }
    C() {
        if (this.disposed)
            return; // Disposed, it shouldn't be updated again
        if (this.J === DIRTY_MAYBE_YES) { // Maybe we are dirty, let's check with our A, to be sure
            this.A.C();
        }
        if (this.J === DIRTY_YES) { // We are dirty, let's H
            this.J = DIRTY_MAYBE_NO; // Trip flag, to be able to tell if we caused ourselves to be dirty again
            this.run();
            if (this.J === DIRTY_MAYBE_NO) { // Not dirty anymore
                this.J = DIRTY_NO;
            }
            else { // Maybe we are still dirty, let's check again
                this.C();
            }
        }
        else { // Not dirty
            this.J = DIRTY_NO;
        }
    }
}
/* EXPORT */
export default Observer;
