// oby@0.54.0/dist/objects/observable.js
/* IMPORT */
import { DIRTY_MAYBE_NO, DIRTY_YES, UNINITIALIZED } from '../constants.js';
import { OBSERVER } from '../context.js';
import Scheduler from './scheduler.sync.js';
import { is, nope } from '../utils.js';
/* MAIN */
class Observable {
    /* CONSTRUCTOR */
    constructor(value, options, parent) {
        this.K = new Set();
        this.value = value;
        if (parent) {
            this.parent = parent;
        }
        if (options?.equals !== undefined) {
            this.equals = options.equals || nope;
        }
    }
    /* API */
    get() {
        if (!this.parent?.disposed) {
            this.parent?.C();
            OBSERVER?.A.L(this);
        }
        return this.value;
    }
    set(value) {
        const equals = this.equals || is;
        const fresh = (this.value === UNINITIALIZED) || !equals(value, this.value);
        if (!fresh)
            return value;
        this.value = value;
        Scheduler.M += 1;
        this.I(DIRTY_YES);
        Scheduler.M -= 1;
        Scheduler.N();
        return value;
    }
    I(J) {
        for (const observer of this.K) {
            if (observer.J !== DIRTY_MAYBE_NO || observer.A.has(this)) { // Maybe this is a potential future dependency we haven't re-read yet
                if (observer.sync) {
                    observer.J = Math.max(observer.J, J);
                    Scheduler.F(observer);
                }
                else {
                    observer.I(J);
                }
            }
        }
    }
    C(fn) {
        const value = fn(this.value);
        return this.set(value);
    }
}
/* EXPORT */
export default Observable;
