// oby@0.54.0/dist/objects/scheduler.sync.js
/* IMPORT */
/* MAIN */
// This ensures that there's only one flushing of the queue happening at the same time
class Scheduler {
    constructor() {
        /* VARIABLES */
        this.A1 = [];
        this.M = 0;
        this.A2 = false;
        /* QUEING API */
        this.N = () => {
            if (this.A2)
                return;
            if (this.M)
                return;
            if (!this.A1.length)
                return;
            try {
                this.A2 = true;
                while (true) {
                    const queue = this.A1;
                    if (!queue.length)
                        break;
                    this.A1 = [];
                    for (let i = 0, l = queue.length; i < l; i++) {
                        queue[i].C();
                    }
                }
            }
            finally {
                this.A2 = false;
            }
        };
        this.E = (fn) => {
            this.M += 1;
            fn();
            this.M -= 1;
            this.N();
        };
        /* SCHEDULING API */
        this.F = (observer) => {
            this.A1.push(observer);
        };
    }
}
/* EXPORT */
export default new Scheduler();
