// oby@14.3.5/dist/methods/for.cache.unkeyed.js
/* IMPORT */
import { OWNER } from '../context.js';
import { lazySetAdd, lazySetDelete } from '../lazy.js';
import cleanup from './cleanup.js';
import get from './get.js';
import memo from './memo.js';
import resolve from './resolve.js';
import suspense from './suspense.js';
import { frozen, readable } from '../objects/callable.js';
import Observable from '../objects/observable.js';
import Root from '../objects/root.js';
import { SYMBOL_CACHED, SYMBOL_SUSPENSE, SYMBOL_UNCACHED } from '../symbols.js';
/* HELPERS */
const DUMMY_INDEX = frozen(-1);
class MappedRoot extends Root {
}
/* MAIN */
//TODO: Optimize this more
class CacheUnkeyed {
    /* CONSTRUCTOR */
    constructor(fn, pooled) {
        /* VARIABLES */
        this.parent = OWNER;
        this.suspense = OWNER.get(SYMBOL_SUSPENSE);
        this.A5 = new Map();
        this.AC = [];
        this.AD = 0;
        /* API */
        this.cleanup = () => {
            let pooled = 0;
            let poolable = Math.max(0, this.pooled ? this.AD - this.AC.length : 0);
            this.A5.forEach(mapped => {
                if (poolable > 0 && pooled++ < poolable) {
                    mapped.suspended?.set(true);
                    this.AC.push(mapped);
                }
                else {
                    mapped.Q(true);
                }
            });
        };
        this.Q = () => {
            if (this.suspense) {
                lazySetDelete(this.parent, 'T', this.T);
            }
            this.A5.forEach(mapped => {
                mapped.Q(true);
            });
            this.AC.forEach(mapped => {
                mapped.Q(true);
            });
        };
        this.map = (values) => {
            const { A5, fn, AB } = this;
            const cacheNext = new Map();
            const results = new Array(values.length);
            const AC = this.AC;
            const pooled = this.pooled;
            let resultsCached = true; // Whether all results are cached, if so this enables an optimization
            let resultsUncached = true; // Whether all results are anew, if so this enables an optimization in Voby
            let leftovers = [];
            if (A5.size) {
                for (let i = 0, l = values.length; i < l; i++) {
                    const value = values[i];
                    const cached = A5.get(value);
                    if (cached) {
                        resultsUncached = false;
                        A5.delete(value);
                        cacheNext.set(value, cached);
                        cached.index?.set(i);
                        results[i] = cached.G; //TSC
                    }
                    else {
                        leftovers.push(i);
                    }
                }
            }
            else {
                leftovers = new Array(results.length);
            }
            outer: for (let i = 0, l = leftovers.length; i < l; i++) {
                const index = leftovers[i] || i;
                const value = values[index];
                const isDuplicate = cacheNext.has(value);
                if (!isDuplicate) {
                    for (const [key, mapped] of A5.entries()) {
                        A5.delete(key);
                        cacheNext.set(value, mapped);
                        mapped.index?.set(index);
                        mapped.value?.set(value);
                        results[index] = mapped.G; //TSC
                        continue outer;
                    }
                }
                resultsCached = false;
                let mapped;
                if (pooled && AC.length) {
                    mapped = AC.pop(); //TSC
                    mapped.index?.set(index);
                    mapped.value?.set(value);
                    mapped.suspended?.set(false);
                    results[index] = mapped.G; //TSC
                }
                else {
                    mapped = new MappedRoot(false);
                    mapped.E(() => {
                        let $index = DUMMY_INDEX;
                        if (AB) {
                            mapped.index = new Observable(index);
                            $index = readable(mapped.index);
                        }
                        const observable = mapped.value = new Observable(value);
                        const suspended = pooled ? new Observable(false) : undefined;
                        const $value = memo(() => get(observable.get())); //TSC
                        const G = results[index] = suspended ? suspense(() => suspended.get(), () => resolve(fn($value, $index))) : resolve(fn($value, $index));
                        mapped.value = observable;
                        mapped.G = G;
                        mapped.suspended = suspended;
                    });
                }
                if (isDuplicate) { // Expensive, not reusable
                    cleanup(() => mapped.Q(true));
                }
                else { // Cheap, reusable
                    cacheNext.set(value, mapped);
                }
            }
            this.AD = Math.max(this.AD, results.length);
            this.cleanup();
            this.A5 = cacheNext;
            if (resultsCached) {
                results[SYMBOL_CACHED] = true;
            }
            if (resultsUncached) {
                results[SYMBOL_UNCACHED] = true;
            }
            return results;
        };
        this.T = () => {
            return [...this.A5.values(), ...this.AC.values()];
        };
        this.fn = fn;
        this.AB = (fn.length > 1);
        this.pooled = pooled;
        if (this.suspense) {
            lazySetAdd(this.parent, 'T', this.T);
        }
    }
}
/* EXPORT */
export default CacheUnkeyed;
