// oby@14.3.5/dist/methods/store.js
/* IMPORT */
import { BATCH, OBSERVER } from '../context.js';
import { lazySetAdd, lazySetDelete, lazySetEach } from '../lazy.js';
import cleanup from './cleanup.js';
import effect from './effect.js';
import isBatching from './is_batching.js';
import isStore from './is_store.js';
import untrack from './untrack.js';
import { readable } from '../objects/callable.js';
import ObservableClass from '../objects/observable.js';
import { SYMBOL_STORE, SYMBOL_STORE_KEYS, SYMBOL_STORE_OBSERVABLE, SYMBOL_STORE_TARGET, SYMBOL_STORE_VALUES, SYMBOL_STORE_UNTRACKED } from '../symbols.js';
import { castArray, is, isArray, isFunction, isObject, noop, nope } from '../utils.js';
/* CLASSES */
class StoreMap extends Map {
    AH(key, value) {
        super.set(key, value);
        return value;
    }
}
class StoreCleanable {
    constructor() {
        this.AE = 0;
    }
    listen() {
        this.AE += 1;
        cleanup(this);
    }
    call() {
        this.AE -= 1;
        if (this.AE)
            return;
        this.Q();
    }
    Q() { }
}
class StoreKeys extends StoreCleanable {
    constructor(parent, observable) {
        super();
        this.parent = parent;
        this.observable = observable;
    }
    Q() {
        this.parent.keys = undefined;
    }
}
class StoreValues extends StoreCleanable {
    constructor(parent, observable) {
        super();
        this.parent = parent;
        this.observable = observable;
    }
    Q() {
        this.parent.values = undefined;
    }
}
class StoreHas extends StoreCleanable {
    constructor(parent, key, observable) {
        super();
        this.parent = parent;
        this.key = key;
        this.observable = observable;
    }
    Q() {
        this.parent.has?.delete(this.key);
    }
}
class StoreProperty extends StoreCleanable {
    constructor(parent, key, observable, AI) {
        super();
        this.parent = parent;
        this.key = key;
        this.observable = observable;
        this.AI = AI;
    }
    Q() {
        this.parent.AJ?.delete(this.key);
    }
}
const StoreListenersRegular = {
    /* VARIABLES */
    AK: 0,
    AL: new Set(),
    AM: new Set(),
    /* API */
    AN: () => {
        const { AL, AM } = StoreListenersRegular;
        const traversed = new Set();
        const traverse = (AI) => {
            if (traversed.has(AI))
                return;
            traversed.add(AI);
            lazySetEach(AI.AO, traverse);
            lazySetEach(AI.AP, listener => {
                AL.add(listener);
            });
        };
        AM.forEach(traverse);
        return () => {
            AL.forEach(listener => {
                listener();
            });
        };
    },
    V: (AI) => {
        StoreListenersRegular.AM.add(AI);
        StoreScheduler.F();
    },
    reset: () => {
        StoreListenersRegular.AL = new Set();
        StoreListenersRegular.AM = new Set();
    }
};
const StoreListenersRoots = {
    /* VARIABLES */
    AK: 0,
    AM: new Map(),
    /* API */
    AN: () => {
        const { AM } = StoreListenersRoots;
        return () => {
            AM.forEach((rootsSet, store) => {
                const T = Array.from(rootsSet);
                lazySetEach(store.AQ, listener => {
                    listener(T);
                });
            });
        };
    },
    V: (store, root) => {
        const T = StoreListenersRoots.AM.get(store) || new Set();
        T.add(root);
        StoreListenersRoots.AM.set(store, T);
        StoreScheduler.F();
    },
    AR: (current, parent, key) => {
        if (!parent.AO) {
            const root = current?.store || untrack(() => parent.store[key]);
            StoreListenersRoots.V(parent, root);
        }
        else {
            const traversed = new Set();
            const traverse = (AI) => {
                if (traversed.has(AI))
                    return;
                traversed.add(AI);
                lazySetEach(AI.AO, parent => {
                    if (!parent.AO) {
                        StoreListenersRoots.V(parent, AI.store);
                    }
                    traverse(parent);
                });
            };
            traverse(current || parent);
        }
    },
    reset: () => {
        StoreListenersRoots.AM = new Map();
    }
};
const StoreScheduler = {
    /* VARIABLES */
    AK: false,
    /* API */
    N: () => {
        const flushRegular = StoreListenersRegular.AN();
        const flushRoots = StoreListenersRoots.AN();
        StoreScheduler.reset();
        flushRegular();
        flushRoots();
    },
    AS: () => {
        if (isBatching()) {
            if (BATCH) {
                BATCH.finally(StoreScheduler.AS);
            }
            else {
                setTimeout(StoreScheduler.AS, 0);
            }
        }
        else {
            StoreScheduler.N();
        }
    },
    reset: () => {
        StoreScheduler.AK = false;
        StoreListenersRegular.reset();
        StoreListenersRoots.reset();
    },
    F: () => {
        if (StoreScheduler.AK)
            return;
        StoreScheduler.AK = true;
        queueMicrotask(StoreScheduler.AS);
    }
};
/* CONSTANTS */
const NODES = new WeakMap();
const SPECIAL_SYMBOLS = new Set([SYMBOL_STORE, SYMBOL_STORE_KEYS, SYMBOL_STORE_OBSERVABLE, SYMBOL_STORE_TARGET, SYMBOL_STORE_VALUES]);
const UNREACTIVE_KEYS = new Set(['__proto__', '__defineGetter__', '__defineSetter__', '__lookupGetter__', '__lookupSetter__', 'prototype', 'constructor', 'hasOwnProperty', 'isPrototypeOf', 'propertyIsEnumerable', 'toLocaleString', 'toSource', 'toString', 'valueOf']);
const STORE_TRAPS = {
    /* API */
    get: (target, key) => {
        if (SPECIAL_SYMBOLS.has(key)) {
            if (key === SYMBOL_STORE)
                return true;
            if (key === SYMBOL_STORE_TARGET)
                return target;
            if (key === SYMBOL_STORE_KEYS) {
                if (isListenable()) {
                    const AI = getNodeExisting(target);
                    AI.keys || (AI.keys = getNodeKeys(AI));
                    AI.keys.listen();
                    AI.keys.observable.get();
                }
                return;
            }
            if (key === SYMBOL_STORE_VALUES) {
                if (isListenable()) {
                    const AI = getNodeExisting(target);
                    AI.values || (AI.values = getNodeValues(AI));
                    AI.values.listen();
                    AI.values.observable.get();
                }
                return;
            }
            if (key === SYMBOL_STORE_OBSERVABLE) {
                return (key) => {
                    key = (typeof key === 'number') ? String(key) : key;
                    const AI = getNodeExisting(target);
                    const getter = AI.AT?.get(key);
                    if (getter)
                        return getter.bind(AI.store);
                    AI.AJ || (AI.AJ = new StoreMap());
                    const value = target[key];
                    const property = AI.AJ.get(key) || AI.AJ.AH(key, getNodeProperty(AI, key, value));
                    const options = AI.equals ? { equals: AI.equals } : undefined;
                    property.observable || (property.observable = getNodeObservable(AI, value, options));
                    const observable = readable(property.observable);
                    return observable;
                };
            }
        }
        if (UNREACTIVE_KEYS.has(key))
            return target[key];
        const AI = getNodeExisting(target);
        const getter = AI.AT?.get(key);
        const value = getter || target[key];
        AI.AJ || (AI.AJ = new StoreMap());
        const listenable = isListenable();
        const proxiable = isProxiable(value);
        const property = listenable || proxiable ? AI.AJ.get(key) || AI.AJ.AH(key, getNodeProperty(AI, key, value)) : undefined;
        if (property?.AI) {
            lazySetAdd(property.AI, 'AO', AI);
        }
        if (property && listenable) {
            const options = AI.equals ? { equals: AI.equals } : undefined;
            property.listen();
            property.observable || (property.observable = getNodeObservable(AI, value, options));
            property.observable.get();
        }
        if (getter) {
            return getter.call(AI.store);
        }
        else {
            if (typeof value === 'function' && value === Array.prototype[key]) {
                return function () {
                    return value.apply(AI.store, arguments);
                };
            }
            return property?.AI?.store || value;
        }
    },
    set: (target, key, value) => {
        value = getTarget(value);
        const AI = getNodeExisting(target);
        const setter = AI.AU?.get(key);
        if (setter) {
            setter.call(AI.store, value);
        }
        else {
            const valuePrev = target[key];
            const hadProperty = !!valuePrev || (key in target);
            const equals = AI.equals || is;
            if (hadProperty && equals(value, valuePrev) && (key !== 'length' || !Array.isArray(target)))
                return true; // Array.prototype.length is special, it gets updated AA this trap is called, we need to special-case it...
            target[key] = value;
            AI.values?.observable.set(0);
            if (!hadProperty) {
                AI.keys?.observable.set(0);
                AI.has?.get(key)?.observable.set(true);
            }
            const property = AI.AJ?.get(key);
            if (property?.AI) {
                lazySetDelete(property.AI, 'AO', AI);
            }
            if (property) {
                property.observable?.set(value);
                property.AI = isProxiable(value) ? NODES.get(value) || getNode(value, AI) : undefined;
            }
            if (property?.AI) {
                lazySetAdd(property.AI, 'AO', AI);
            }
            if (StoreListenersRoots.AK) {
                StoreListenersRoots.AR(property?.AI, AI, key);
            }
            if (StoreListenersRegular.AK) {
                StoreListenersRegular.V(AI);
            }
        }
        return true;
    },
    deleteProperty: (target, key) => {
        const hasProperty = (key in target);
        if (!hasProperty)
            return true;
        const deleted = Reflect.deleteProperty(target, key);
        if (!deleted)
            return false;
        const AI = getNodeExisting(target);
        AI.keys?.observable.set(0);
        AI.values?.observable.set(0);
        AI.has?.get(key)?.observable.set(false);
        const property = AI.AJ?.get(key);
        if (StoreListenersRoots.AK) {
            StoreListenersRoots.AR(property?.AI, AI, key);
        }
        if (property?.AI) {
            lazySetDelete(property.AI, 'AO', AI);
        }
        if (property) {
            property.observable?.set(undefined);
            property.AI = undefined;
        }
        if (StoreListenersRegular.AK) {
            StoreListenersRegular.V(AI);
        }
        return true;
    },
    defineProperty: (target, key, descriptor) => {
        const AI = getNodeExisting(target);
        const equals = AI.equals || is;
        const hadProperty = (key in target);
        const descriptorPrev = Reflect.getOwnPropertyDescriptor(target, key);
        if ('value' in descriptor && isStore(descriptor.value)) {
            descriptor = { ...descriptor, value: getTarget(descriptor.value) };
        }
        if (descriptorPrev && isEqualDescriptor(descriptorPrev, descriptor, equals))
            return true;
        const defined = Reflect.defineProperty(target, key, descriptor);
        if (!defined)
            return false;
        if (!descriptor.get) {
            AI.AT?.delete(key);
        }
        else if (descriptor.get) {
            AI.AT || (AI.AT = new StoreMap());
            AI.AT.set(key, descriptor.get);
        }
        if (!descriptor.set) {
            AI.AU?.delete(key);
        }
        else if (descriptor.set) {
            AI.AU || (AI.AU = new StoreMap());
            AI.AU.set(key, descriptor.set);
        }
        if (hadProperty !== !!descriptor.enumerable) {
            AI.keys?.observable.set(0);
        }
        AI.has?.get(key)?.observable.set(true);
        const property = AI.AJ?.get(key);
        if (StoreListenersRoots.AK) {
            StoreListenersRoots.AR(property?.AI, AI, key);
        }
        if (property?.AI) {
            lazySetDelete(property.AI, 'AO', AI);
        }
        if (property) {
            if ('get' in descriptor) {
                property.observable?.set(descriptor.get);
                property.AI = undefined;
            }
            else {
                const value = descriptor.value;
                property.observable?.set(value);
                property.AI = isProxiable(value) ? NODES.get(value) || getNode(value, AI) : undefined;
            }
        }
        if (property?.AI) {
            lazySetAdd(property.AI, 'AO', AI);
        }
        if (StoreListenersRoots.AK) {
            StoreListenersRoots.AR(property?.AI, AI, key);
        }
        if (StoreListenersRegular.AK) {
            StoreListenersRegular.V(AI);
        }
        return true;
    },
    has: (target, key) => {
        if (key === SYMBOL_STORE)
            return true;
        if (key === SYMBOL_STORE_TARGET)
            return true;
        const value = (key in target);
        if (isListenable()) {
            const AI = getNodeExisting(target);
            AI.has || (AI.has = new StoreMap());
            const has = AI.has.get(key) || AI.has.AH(key, getNodeHas(AI, key, value));
            has.listen();
            has.observable.get();
        }
        return value;
    },
    ownKeys: (target) => {
        const keys = Reflect.ownKeys(target);
        if (isListenable()) {
            const AI = getNodeExisting(target);
            AI.keys || (AI.keys = getNodeKeys(AI));
            AI.keys.listen();
            AI.keys.observable.get();
        }
        return keys;
    }
};
const STORE_UNTRACK_TRAPS = {
    /* API */
    has: (target, key) => {
        if (key === SYMBOL_STORE_UNTRACKED)
            return true;
        return (key in target);
    }
};
/* HELPERS */
const getNode = (value, parent, equals) => {
    const store = new Proxy(value, STORE_TRAPS);
    const gettersAndSetters = getGettersAndSetters(value);
    const AI = { AO: parent, store };
    if (gettersAndSetters) {
        const { AT, AU } = gettersAndSetters;
        if (AT)
            AI.AT = AT;
        if (AU)
            AI.AU = AU;
    }
    if (equals === false) {
        AI.equals = nope;
    }
    else if (equals) {
        AI.equals = equals;
    }
    else if (parent?.equals) {
        AI.equals = parent.equals;
    }
    NODES.set(value, AI);
    return AI;
};
const getNodeExisting = (value) => {
    const AI = NODES.get(value);
    if (!AI)
        throw new Error('Impossible');
    return AI;
};
const getNodeFromStore = (store) => {
    return getNodeExisting(getTarget(store));
};
const getNodeKeys = (AI) => {
    const observable = getNodeObservable(AI, 0, { equals: false });
    const keys = new StoreKeys(AI, observable);
    return keys;
};
const getNodeValues = (AI) => {
    const observable = getNodeObservable(AI, 0, { equals: false });
    const values = new StoreValues(AI, observable);
    return values;
};
const getNodeHas = (AI, key, value) => {
    const observable = getNodeObservable(AI, value);
    const has = new StoreHas(AI, key, observable);
    return has;
};
const getNodeObservable = (AI, value, options) => {
    return new ObservableClass(value, options);
};
const getNodeProperty = (AI, key, value) => {
    const observable = undefined;
    const propertyNode = isProxiable(value) ? NODES.get(value) || getNode(value, AI) : undefined;
    const property = new StoreProperty(AI, key, observable, propertyNode);
    AI.AJ || (AI.AJ = new StoreMap());
    AI.AJ.set(key, property);
    return property;
};
const getGettersAndSetters = (value) => {
    if (isArray(value))
        return;
    let AT;
    let AU;
    const keys = Object.keys(value);
    for (let i = 0, l = keys.length; i < l; i++) {
        const key = keys[i];
        const descriptor = Object.getOwnPropertyDescriptor(value, key);
        if (!descriptor)
            continue;
        const { get, set } = descriptor;
        if (get) {
            AT || (AT = new StoreMap());
            AT.set(key, get);
        }
        if (set) {
            AU || (AU = new StoreMap());
            AU.set(key, set);
        }
    }
    if (!AT && !AU)
        return;
    return { AT, AU };
};
const getStore = (value, options) => {
    if (isStore(value))
        return value;
    const AI = NODES.get(value) || getNode(value, undefined, options?.equals);
    return AI.store;
};
const getTarget = (value) => {
    if (isStore(value))
        return value[SYMBOL_STORE_TARGET];
    return value;
};
const getUntracked = (value) => {
    if (!isObject(value))
        return value;
    if (isUntracked(value))
        return value;
    return new Proxy(value, STORE_UNTRACK_TRAPS);
};
const isEqualDescriptor = (a, b, equals) => {
    return (!!a.configurable === !!b.configurable && !!a.enumerable === !!b.enumerable && !!a.writable === !!b.writable && equals(a.value, b.value) && a.get === b.get && a.set === b.set);
};
const isListenable = () => {
    return !!OBSERVER;
};
const isProxiable = (value) => {
    if (value === null || typeof value !== 'object')
        return false;
    if (SYMBOL_STORE in value)
        return true;
    if (SYMBOL_STORE_UNTRACKED in value)
        return false;
    if (isArray(value))
        return true;
    const prototype = Object.getPrototypeOf(value);
    if (prototype === null)
        return true;
    return (Object.getPrototypeOf(prototype) === null);
};
const isUntracked = (value) => {
    if (value === null || typeof value !== 'object')
        return false;
    return (SYMBOL_STORE_UNTRACKED in value);
};
/* MAIN */
//TODO: Maybe have the "on" method trigger immediately too like "$.on", or maybe the other way around, which seems more flexible
//TODO: Explore converting target values back to numbers (the Proxy always receives strings) whenever possible, to save memory
//TODO: Implement "_onRoots" better, perhaps provding string paths instead, which should be more powerful
const store = (value, options) => {
    if (!isObject(value))
        return value;
    if (isUntracked(value))
        return value;
    return getStore(value, options);
};
/* UTILITIES */
store.on = (target, listener) => {
    /* VARIABLES */
    const targets = isStore(target) ? [target] : castArray(target);
    const selectors = targets.filter(isFunction);
    const AM = targets.filter(isStore).map(getNodeFromStore);
    /* ON */
    StoreListenersRegular.AK += 1;
    const disposers = selectors.map(selector => {
        let inited = false;
        return effect(() => {
            if (inited) {
                StoreListenersRegular.AL.add(listener);
                StoreScheduler.F();
            }
            inited = true;
            selector();
        }, { suspense: false, sync: true });
    });
    AM.forEach(AI => {
        lazySetAdd(AI, 'AP', listener);
    });
    /* OFF */
    return () => {
        StoreListenersRegular.AK -= 1;
        disposers.forEach(disposer => {
            disposer();
        });
        AM.forEach(AI => {
            lazySetDelete(AI, 'AP', listener);
        });
    };
};
store._onRoots = (target, listener) => {
    if (!isStore(target))
        return noop;
    const AI = getNodeFromStore(target);
    if (AI.AO)
        throw new Error('Only top-level stores are supported');
    /* ON */
    StoreListenersRoots.AK += 1;
    lazySetAdd(AI, 'AQ', listener);
    /* OFF */
    return () => {
        StoreListenersRoots.AK -= 1;
        lazySetDelete(AI, 'AQ', listener);
    };
};
store.reconcile = (() => {
    //TODO: Support AT, AU and symbols (symbols could be supported with Reflect.ownKeys, but that's like 2x slower)
    const getType = (value) => {
        if (isArray(value))
            return 1;
        if (isProxiable(value))
            return 2;
        return 0;
    };
    const reconcileOuter = (prev, next) => {
        const uprev = getTarget(prev);
        const unext = getTarget(next);
        reconcileInner(prev, next);
        const prevType = getType(uprev);
        const nextType = getType(unext);
        if (prevType === 1 || nextType === 1) {
            prev.length = next.length;
        }
        return prev;
    };
    const reconcileInner = (prev, next) => {
        const uprev = getTarget(prev);
        const unext = getTarget(next);
        const prevKeys = Object.keys(uprev);
        const nextKeys = Object.keys(unext);
        for (let i = 0, l = nextKeys.length; i < l; i++) {
            const key = nextKeys[i];
            const prevValue = uprev[key];
            const nextValue = unext[key];
            if (!is(prevValue, nextValue)) {
                const prevType = getType(prevValue);
                const nextType = getType(nextValue);
                if (prevType && prevType === nextType) {
                    reconcileInner(prev[key], nextValue);
                    if (prevType === 1) {
                        prev[key].length = nextValue.length;
                    }
                }
                else {
                    prev[key] = nextValue;
                }
            }
            else if (prevValue === undefined && !(key in uprev)) {
                prev[key] = undefined;
            }
        }
        for (let i = 0, l = prevKeys.length; i < l; i++) {
            const key = prevKeys[i];
            if (!(key in unext)) {
                delete prev[key];
            }
        }
        return prev;
    };
    const reconcile = (prev, next) => {
        return untrack(() => {
            return reconcileOuter(prev, next);
        });
    };
    return reconcile;
})();
store.untrack = (value) => {
    return getUntracked(value);
};
store.unwrap = (value) => {
    return getTarget(value);
};
/* EXPORT */
export default store;
